<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вопрос 1:</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Вопрос-2:" data-toc-modified-id="Вопрос-2:-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Вопрос 2:</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Вопрос-3:" data-toc-modified-id="Вопрос-3:-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Вопрос 3:</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Вопрос-4:" data-toc-modified-id="Вопрос-4:-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вопрос 4:</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

# Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.



# Описание данных
 - children — количество детей в семье
 - days_employed — общий трудовой стаж в днях
 - dob_years — возраст клиента в годах
 - education — уровень образования клиента
 - education_id — идентификатор уровня образования
 - family_status — семейное положение
 - family_status_id — идентификатор семейного положения
 - gender — пол клиента
 - income_type — тип занятости
 - debt — имел ли задолженность по возврату кредитов
 - total_income — ежемесячный доход
 - purpose — цель получения кредита

## Обзор данных

In [73]:
#импортируем библиотеку Pandas
import pandas as pd

In [74]:
# загружаем данные
data = pd.read_csv('/datasets/data.csv')
data.info() #выводим общую информацию


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [75]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [76]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

<h2> Вывод <a class="tocSkip"> </h2> Датафрейм содержит 12 колонок и 21525 строк. В двух столбцах - 'days employed' и 'total_income' имеются пропуски.

### Заполнение пропусков

In [77]:
data.isna().sum() #смотрим общую картину пропусков в данных


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [78]:
data.groupby('income_type')['total_income'].count() #основные категории клиентов

income_type
безработный            2
в декрете              1
госслужащий         1312
компаньон           4577
пенсионер           3443
предприниматель        1
сотрудник          10014
студент                1
Name: total_income, dtype: int64

In [79]:
percent_days_employed = ((len(data[data['days_employed'].isna()])) / len(data['days_employed'])) # считаем процент пропусков
f'{percent_days_employed :.0%}' 

'10%'

In [80]:
percent_total_income = ((len(data[data['total_income'].isna()])) / len(data['total_income'])) # считаем процент пропусков
f'{percent_total_income :.0%}'

'10%'

In [81]:
def income_type_isna(type): #функция для подсчета пропущенных значений по категориям клиентов 'income_type'
    return data.loc[data['income_type'] == type].isna().sum()
def income_type_median(type): #функция для подсчета медианы количественных показателей по категориям клиентов 'income_type'
    return data[data['income_type'] == type].sort_values(by = 'income_type').median()

In [82]:
income_type_isna('госслужащий') 

children              0
days_employed       147
dob_years             0
education             0
education_id          0
family_status         0
family_status_id      0
gender                0
income_type           0
debt                  0
total_income        147
purpose               0
dtype: int64

In [83]:
income_type_median('госслужащий')

children                 0.000000
days_employed        -2689.368353
dob_years               40.000000
education_id             1.000000
family_status_id         0.000000
debt                     0.000000
total_income        150447.935283
dtype: float64

In [84]:
income_type_isna('компаньон')

children              0
days_employed       508
dob_years             0
education             0
education_id          0
family_status         0
family_status_id      0
gender                0
income_type           0
debt                  0
total_income        508
purpose               0
dtype: int64

In [85]:
income_type_median('компаньон')

children                 0.000000
days_employed        -1547.382223
dob_years               39.000000
education_id             1.000000
family_status_id         0.000000
debt                     0.000000
total_income        172357.950966
dtype: float64

In [86]:
income_type_isna('пенсионер')

children              0
days_employed       413
dob_years             0
education             0
education_id          0
family_status         0
family_status_id      0
gender                0
income_type           0
debt                  0
total_income        413
purpose               0
dtype: int64

In [87]:
income_type_median('пенсионер')

children                 0.000000
days_employed       365213.306266
dob_years               60.000000
education_id             1.000000
family_status_id         0.000000
debt                     0.000000
total_income        118514.486412
dtype: float64

In [88]:
income_type_isna('сотрудник')

children               0
days_employed       1105
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        1105
purpose                0
dtype: int64

In [89]:
income_type_median('сотрудник')

children                 0.000000
days_employed        -1574.202821
dob_years               39.000000
education_id             1.000000
family_status_id         0.000000
debt                     0.000000
total_income        142594.396847
dtype: float64

In [90]:
days_employed_median = data['days_employed'].sort_values().median() #медианное значение столбца 'days_employed'
days_employed_median

-1203.369528770489

In [91]:
total_income_mediane = data['total_income'].sort_values().median() #медианное значение столбца total_income
total_income_mediane

145017.93753253992

Изучив показатели медиан по всем категориям принято решение заполнить пропуски в столбце 'total_income' общим показателем - total_income_mediane.

In [92]:
data['total_income'] = data['total_income'].fillna(total_income_mediane) #заменяем пропуски на медианные значения

In [93]:
data.isna().sum() #проверяем наличие пропущенных значений для контроля

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

ВЫВОДЫ:
1. Найдены пропущенные значения в двух столбцах - 'days_employed' и 'total_income', тип данных float64.
2. Доля пропусков в столбцах - 'days_employed' и 'total_income' составляет приблизительно 10%.
3. Вероятнее всего причина пропусков в технической ошибке при выгрузке данных, так как количество пропусков в обоих столбцах совпадает и пропуски совпадают.
4. Больше всего пропущенных значений в категориях: 'госслужащий', 'компаньон', 'пенсионер' и 'сотрудник'. Данные из этих категорий быля взяты для более чщательного исследования. После рассмотрения данных по категорям 'income_type' было принято решение замены пропущенных значений в столбце 'days_employed' на значение медианы по категории 'сотрудник', так как это самая многочисленная категория и ее медиана сопоставима с подобным показателем следующей по численности категории - 'компаньон'. Изучив показатели медиан по всем категориям принято решение заполнить пропуски в столбце 'total_income' общим показателем - total_income_mediane, так как данное значение незначительно отличается от медианы осноных категорий.

### Проверка данных на аномалии и исправления.

In [94]:
data.groupby('children')['purpose'].count() #изучаем на аномалии столбец 'children'

children
-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: purpose, dtype: int64

In [95]:
data.loc[data['children'] == -1] #так не бывает

children  days_employed  dob_years            education  education_id  \
291          -1   -4417.703588         46              среднее             1   
705          -1    -902.084528         50              среднее             1   
742          -1   -3174.456205         57              среднее             1   
800          -1  349987.852217         54              среднее             1   
941          -1            NaN         57              Среднее             1   
1363         -1   -1195.264956         55              СРЕДНЕЕ             1   
1929         -1   -1461.303336         38              среднее             1   
2073         -1   -2539.761232         42              среднее             1   
3814         -1   -3045.290443         26              Среднее             1   
4201         -1    -901.101738         41              среднее             1   
4402         -1  398001.302888         64              СРЕДНЕЕ             1   
4542         -1   -1811.899756         32              среднее             1   
5272         -1   -1143.485347         46              среднее             1   
6013         -1   -1361.258696         46               высшее             0   
6385         -1  370215.476226         48              среднее             1   
7190         -1   -5928.202068         34              СРЕДНЕЕ             1   
7278         -1    -526.318451         51              среднее             1   
7615         -1            NaN         35              среднее             1   
7685         -1   -3237.360455         53              среднее             1   
8229         -1   -1803.441074         54               высшее             0   
8251         -1   -3113.998449         31              среднее             1   
8561         -1    -195.479496         31              среднее             1   
9565         -1   -2896.629686         34               высшее             0   
9574         -1   -2710.419901         28               высшее             0   
10177        -1   -1743.604011         37              среднее             1   
10365        -1  345774.125957         63               ВЫСШЕЕ             0   
10918        -1  340499.039342         50              среднее             1   
11109        -1   -3438.463024         59              среднее             1   
11268        -1   -1048.782203         30               высшее             0   
11617        -1    -370.827130         27              среднее             1   
12304        -1   -8493.101252         61  неоконченное высшее             2   
13786        -1            NaN         42              среднее             1   
14357        -1    -268.337037         23              среднее             1   
15147        -1   -9851.184337         44              среднее             1   
15435        -1   -3614.220232         40               высшее             0   
16129        -1    -457.861760         33              среднее             1   
16265        -1   -2802.218127         40               высшее             0   
17095        -1   -2809.693200         34              среднее             1   
17429        -1    -895.379738         37              среднее             1   
17657        -1   -4571.957475         41              среднее             1   
18219        -1   -3575.215641         33              среднее             1   
19100        -1    -617.246968         38              среднее             1   
19366        -1    -372.034749         43              СРЕДНЕЕ             1   
19417        -1  350340.760224         28              среднее             1   
20393        -1  355157.107212         69              среднее             1   
20717        -1    -661.822321         32               высшее             0   
21140        -1   -1422.668059         44              среднее             1   

               family_status  family_status_id gender  income_type  debt  \
291         гражданский брак                 1      F    сотрудник     0   
705          ж

In [96]:
data.loc[data['children'] == 20] #так тоже врядли

children  days_employed  dob_years education  education_id  \
606          20    -880.221113         21   среднее             1   
720          20    -855.595512         44   среднее             1   
1074         20   -3310.411598         56   среднее             1   
2510         20   -2714.161249         59    высшее             0   
2941         20   -2161.591519          0   среднее             1   
...         ...            ...        ...       ...           ...   
21008        20   -1240.257910         40   среднее             1   
21325        20    -601.174883         37   среднее             1   
21390        20            NaN         53   среднее             1   
21404        20    -494.788448         52   среднее             1   
21491        20    -173.954460         27   среднее             1   

         family_status  family_status_id gender income_type  debt  \
606    женат / замужем                 0      M   компаньон     0   
720    женат / замужем                 0      F   компаньон     0   
1074   женат / замужем                 0      F   сотрудник     1   
2510    вдовец / вдова                 2      F   сотрудник     0   
2941   женат / замужем                 0      F   сотрудник     0   
...                ...               ...    ...         ...   ...   
21008  женат / замужем                 0      F   сотрудник     1   
21325  женат / замужем                 0      F   компаньон     0   
21390  женат / замужем                 0      M   компаньон     0   
21404  женат / замужем                 0      M   компаньон     0   
21491  женат / замужем                 0      F   сотрудник     0   

        total_income                                purpose  
606    145334.865002                          покупка жилья  
720    112998.738649                   покупка недвижимости  
1074   229518.537004                  получение образования  
2510   264474.835577  операции с коммерческой недвижимостью  
2941   199739.941398                  на покупку автомобиля  
...              ...                                    ...  
21008  133524.010303                        свой автомобиль  
21325  102986.065978                 профильное образование  
21390  145017.937533             покупка жилой недвижимости  
21404  156629.683642        операции со своей недвижимостью  
21491  155520.899688     на покупку подержанного автомобиля  

[76 rows x 12 columns]

In [97]:
data['children'] = data['children'].abs() #приводим все значения к положительным

In [98]:
data.loc[data['children'] == 20, 'children'] = 2 #заменили 20 на 2

В столбце 'children' аномальными являются значения -1 и 20 и их достаточно много. Изучив выборку по этим значениям ничего необчного не выявлено. Вероятнее всего при внесении данных сработал человеческий фактор, так как значения с одним и двумя ребенком самые многочисленные после значения с 0. Значение -1 меняем на 1, а вместо 20 ставим 2. 

In [99]:
data['days_employed'].describe() #проверка на аномалии столбца 'days_employed'

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [100]:
data.pivot_table(index=['income_type'], values = ['days_employed'], aggfunc = ['min', 'mean', 'max', 'median'])

min           mean            max         median
                 days_employed  days_employed  days_employed  days_employed
income_type                                                                
безработный      337524.466835  366413.652744  395302.838654  366413.652744
в декрете         -3296.759962   -3296.759962   -3296.759962   -3296.759962
госслужащий      -15193.032201   -3399.896902     -39.954170   -2689.368353
компаньон        -17615.563266   -2111.524398     -30.195337   -1547.382223
пенсионер        328728.720605  365003.491245  401755.400475  365213.306266
предприниматель    -520.848083    -520.848083    -520.848083    -520.848083
сотрудник        -18388.949901   -2326.499216     -24.141633   -1574.202821
студент            -578.751554    -578.751554    -578.751554    -578.751554

In [101]:
data['days_employed'] = data['days_employed'].abs() #приводим значения к положительным величинам

In [102]:
days_employed_median = data['days_employed'].sort_values().median() #медианное значение столбца 'days_employed'
days_employed_median

2194.220566878695

In [103]:
data['days_employed'] = data['days_employed'].fillna(days_employed_median) #заменяем пропуски на медианные значения

После проверки аномалий в солбце 'days_employed' по категориям 'income_type' очивидны следующие аномалии:
1. Имеются отрицательные значения. В целом это похоже на техническую ошибку, так как в категориях 'в декрете', 'госслужащий', 'компаньон', 'предприниматель', 'сотрудник' и 'студент' нет положительных значений. Все значения в этом стобце приведем в положительные, в этом случае значения буду логичными.
2. Значения'безработные' нереальны. Так как их ничтожно мало по сравнением с общим количеством, можно их либо оставить, как есть, либо удалить. Они не влияют на общую картину. 
3. Значения 'пенсионеров' нужно уточнить у разработчиков, так как средний показатель так же не реален.

In [104]:
#проверка столбца 'dob_years'
data.pivot_table(index=['income_type'], values = ['dob_years'], aggfunc = ['min', 'mean', 'max', 'median'])

min       mean       max    median
                dob_years  dob_years dob_years dob_years
income_type                                             
безработный            31  38.000000        45      38.0
в декрете              39  39.000000        39      39.0
госслужащий             0  40.636737        75      40.0
компаньон               0  39.697542        74      39.0
пенсионер               0  59.063019        74      60.0
предприниматель        27  42.500000        58      42.5
сотрудник               0  39.821027        74      39.0
студент                22  22.000000        22      22.0

In [105]:
def dob_years_zero(types): #функция для подсчета нулевых значений в столбце 'dob_years'
    dob = data[data['dob_years'] == 0]
    dob = dob[dob['income_type'] == types]
    return dob['dob_years'].count()



In [106]:
dob_years_zero('госслужащий')

6

In [107]:
dob_years_zero('компаньон')

20

In [108]:
dob_years_zero('пенсионер')

20

In [109]:
dob_years_zero('сотрудник')

55

Меняем нулевые значения на медианные, согласно типа ('income_type')

In [110]:
clerc = (data['dob_years'] == 0) & (data['income_type'] == 'госслужащий')
data.loc[clerc, 'dob_years'] = 40
partner = (data['dob_years'] == 0) & (data['income_type'] == 'компаньон')
data.loc[partner, 'dob_years'] = 39
retire = (data['dob_years'] == 0) & (data['income_type'] == 'пенсионер')
data.loc[retire, 'dob_years'] = 60
employee = (data['dob_years'] == 0) & (data['income_type'] == 'сотрудник')
data.loc[employee, 'dob_years'] = 39


In [111]:
#проверка столбца 'dob_years'
data.pivot_table(index=['income_type'], values = ['dob_years'], aggfunc = ['min', 'mean', 'max', 'median'])

min       mean       max    median
                dob_years  dob_years dob_years dob_years
income_type                                             
безработный            31  38.000000        45      38.0
в декрете              39  39.000000        39      39.0
госслужащий            19  40.801234        75      40.0
компаньон              19  39.850934        74      39.0
пенсионер              22  59.374222        74      60.0
предприниматель        27  42.500000        58      42.5
сотрудник              19  40.013940        74      39.0
студент                22  22.000000        22      22.0

In [112]:
data['gender'].unique() #проверка столбца 'gender'

array(['F', 'M', 'XNA'], dtype=object)

In [113]:
data.groupby('gender')['gender'].count()

gender
F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [114]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

In [115]:
# удаляем пользователя с полом XNA
data = data.drop(data[data['gender'] == 'XNA'].index)


ВЫВОДЫ:
Аномалии встретились в нескольких столбах:
- 'children' аномальными являются значения -1 и 20 и их достаточно много. Изучив выборку по этим значениям ничего необчного не выявлено. Вероятнее всего при внесении данных сработал человеческий фактор, так как значения с одним и двумя ребенком самые многочисленные после значения с 0. Значение -1 нужно меняем на 1, а вместо 20 ставим 2.
- 'days_employed' имеются отрицательные значения. В целом это похоже на техническую ошибку, так как в категориях 'в декрете', 'госслужащий', 'компаньон', 'предприниматель', 'сотрудник' и 'студент' нет положительных значений. Все значения в этом стобце привели в положительные, в этом случае значения буду логичными.
   Значения'безработные' нереальны. Так как их ничтожно мало по сравнением с общим количеством, можно их либо оставить, как есть, либо удалить. Они не влияют на общую картину. 
   Значения 'пенсионеров' нужно уточнить у разработчиков, так как средний показатель так же не реален.
- 'dob_years' в некоторых значениях поставлены нули. Данная аномалия исправлена медианными значениями в соответствии с типом.
- 'gender' есть одно значение 'XNA', непонятно, что означает, удалим его.

### Изменение типов данных.

In [116]:
data['total_income'] = data['total_income'].astype('int') #изменяем значения в 'total_income' с вещественных на целочисленные

In [117]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Удаление дубликатов.

In [118]:
data.duplicated().sum()

54

In [119]:
data[data.duplicated()]

children  days_employed  dob_years education  education_id  \
2849          0    2194.220567         41   среднее             1   
4182          1    2194.220567         34    ВЫСШЕЕ             0   
4851          0    2194.220567         60   среднее             1   
5557          0    2194.220567         58   среднее             1   
7808          0    2194.220567         57   среднее             1   
8583          0    2194.220567         58    высшее             0   
9238          2    2194.220567         34   среднее             1   
9528          0    2194.220567         66   среднее             1   
9627          0    2194.220567         56   среднее             1   
10462         0    2194.220567         62   среднее             1   
10697         0    2194.220567         40   среднее             1   
10864         0    2194.220567         62   среднее             1   
10994         0    2194.220567         62   среднее             1   
11791         0    2194.220567         47   среднее             1   
12373         0    2194.220567         58   среднее             1   
12375         1    2194.220567         37   среднее             1   
12736         0    2194.220567         59   среднее             1   
13025         1    2194.220567         44   среднее             1   
13639         0    2194.220567         64   среднее             1   
13773         0    2194.220567         35   среднее             1   
13878         1    2194.220567         31   среднее             1   
13942         0    2194.220567         44   среднее             1   
14432         2    2194.220567         36    высшее             0   
14832         0    2194.220567         50   среднее             1   
15091         0    2194.220567         58   среднее             1   
15188         0    2194.220567         60   среднее             1   
15273         0    2194.220567         57   среднее             1   
16176         0    2194.220567         47   среднее             1   
16378         0    2194.220567         46   среднее             1   
16902         2    2194.220567         39   среднее             1   
16904         1    2194.220567         32    высшее             0   
17379         0    2194.220567         54    высшее             0   
17755         1    2194.220567         43   среднее             1   
17774         1    2194.220567         40   среднее             1   
18328         0    2194.220567         29    высшее             0   
18349         1    2194.220567         30    высшее             0   
18428         0    2194.220567         64   среднее             1   
18521         0    2194.220567         56   среднее             1   
18563         0    2194.220567         54   среднее             1   
18755         0    2194.220567         58   среднее             1   
19041         0    2194.220567         56   среднее             1   
19184         0    2194.220567         46   среднее             1   
19321         0    2194.220567         23   среднее             1   
19387         0    2194.220567         38    высшее             0   
19688         0    2194.220567         61   среднее             1   
19832         0    2194.220567         48   среднее             1   
19946         0    2194.220567         57   среднее             1   
20116         0    2194.220567         57   среднее             1   
20165         0    2194.220567         42   среднее             1   
20702         0    2194.220567         64   среднее             1   
21032         0    2194.220567         60   среднее             1   
21132         0    2194.220567         47   среднее             1   
21281         1    2194.220567         30    высшее             0   
21415         0    2194.220567         54   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
2849         женат / замужем                 0      F    сотрудник     0   
4182        гражданский брак                 1      F    со

In [120]:
data = data.drop_duplicates().reset_index(drop = True) #удаляем явные дубликаты

<div class="alert alert-success">
<h2> Комментарий ревьюера ✔️ <a class="tocSkip"> </h2>

Все круто. Сбросить индексы - отличное решение, на мой взгляд, так лучше.

In [121]:
data.duplicated().sum()

0

In [122]:
data.groupby('education')['education_id'].value_counts() #проверка неявных дубликатов

education            education_id
ВЫСШЕЕ               0                 273
Высшее               0                 268
НАЧАЛЬНОЕ            3                  17
НЕОКОНЧЕННОЕ ВЫСШЕЕ  2                  29
Начальное            3                  15
Неоконченное высшее  2                  47
СРЕДНЕЕ              1                 772
Среднее              1                 711
УЧЕНАЯ СТЕПЕНЬ       4                   1
Ученая степень       4                   1
высшее               0                4710
начальное            3                 250
неоконченное высшее  2                 667
среднее              1               13705
ученая степень       4                   4
Name: education_id, dtype: int64

In [123]:
data['education'] = data['education'].str.lower() #приводим значения в столбце 'education' к нижнему регистру

In [124]:
data.groupby('education')['education_id'].value_counts() #дубликатов нет

education            education_id
высшее               0                5251
начальное            3                 282
неоконченное высшее  2                 743
среднее              1               15188
ученая степень       4                   6
Name: education_id, dtype: int64

In [125]:
data.groupby('family_status')['family_status_id'].value_counts() #проверка неявных дубликатов 

family_status          family_status_id
Не женат / не замужем  4                    2810
в разводе              3                    1195
вдовец / вдова         2                     959
гражданский брак       1                    4162
женат / замужем        0                   12344
Name: family_status_id, dtype: int64

In [126]:
data['family_status'] = data['family_status'].str.lower() #приводим значения в столбце 'family_status' к нижнему регистру

In [127]:
data.groupby('family_status')['family_status_id'].value_counts() #дубликатов нет

family_status          family_status_id
в разводе              3                    1195
вдовец / вдова         2                     959
гражданский брак       1                    4162
женат / замужем        0                   12344
не женат / не замужем  4                    2810
Name: family_status_id, dtype: int64

In [128]:
data.groupby('income_type')['debt'].unique() #в столбце 'income_type' дубликатов нет.

income_type
безработный        [1, 0]
в декрете             [1]
госслужащий        [0, 1]
компаньон          [0, 1]
пенсионер          [0, 1]
предприниматель       [0]
сотрудник          [0, 1]
студент               [0]
Name: debt, dtype: object

In [129]:
data.groupby('purpose')['purpose'].unique()

purpose
автомобили                                                            [автомобили]
автомобиль                                                            [автомобиль]
высшее образование                                            [высшее образование]
дополнительное образование                            [дополнительное образование]
жилье                                                                      [жилье]
заняться высшим образованием                        [заняться высшим образованием]
заняться образованием                                      [заняться образованием]
на покупку автомобиля                                      [на покупку автомобиля]
на покупку подержанного автомобиля            [на покупку подержанного автомобиля]
на покупку своего автомобиля                        [на покупку своего автомобиля]
на проведение свадьбы                                      [на проведение свадьбы]
недвижимость                                                        [недвижимос

In [130]:
data['purpose'] = data['purpose'].str.lower() #приводим значения в столбце 'purpose' к нижнему регистру

Разбиваем по категориям: 'недвижимость', 'автомобиль', 'образование', 'свадьба'

In [131]:

def new_purpose(lemmas):     
        if 'авто' in lemmas:
            return 'автомобиль'
        if 'образов' in lemmas:
            return 'образование'
        if 'свад' in lemmas:
            return 'свадьба'
        if 'недвиж' in lemmas:
            return 'недвижимость'
        if 'жиль' in lemmas:
            return 'недвижимость'

new_purpose(data['purpose'])
data.head()
data['new_purpose'] = data['purpose'].apply(new_purpose)
data['new_purpose'].unique()

array(['недвижимость', 'автомобиль', 'образование', 'свадьба'],
      dtype=object)

ВЫВОД:
Дубликаты в основном связаны с человеческим фактором.


### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [134]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose   new_purpose  
0               покупка жилья  недвижимость  
1     приобретение автомобиля    автомобиль  
2               покупка жилья  недвижимость  
3  дополнительное образование   образование  
4             сыграть свадьбу       свадьба

In [135]:
data_education = data[['education', 'education_id']] #создание нового DataFrame
data_education = data_education.drop_duplicates().sort_values(by = 'education_id').reset_index(drop = True)
data_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [136]:
data_family_status = data[['family_status', 'family_status_id']] #создание нового DataFrame
data_family_status = data_family_status.drop_duplicates().sort_values(by = 'family_status_id').reset_index(drop = True)
data_family_status


family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

In [137]:
data = data[['children', 'days_employed', 'dob_years', 'education_id',
       'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose']]
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Категоризация дохода.

In [139]:
def total_income_category(data):
    income = data['total_income']
    if income <= 30000:
        return 'E'
    if 30001 <= income <= 50000:
        return 'D'
    if 50001 <= income <= 200000:
        return 'C'
    if 200001 <= income <= 1000000:
        return 'B'
    if income >= 1000001:
        return 'A'
    
data['total_income_category'] = data.apply(total_income_category, axis = 1)
data.head()
    
    
    

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Категоризация целей кредита.

In [140]:
def purpose_category(data):
    purpose = data['purpose']
    if purpose == 'автомобиль':
        return 'оперции с автомобилем'
    if purpose == 'недвижимость':
        return 'операции с недвижимостью'
    if purpose == 'свадьба':
        return 'проведение свадьбы'
    if purpose == 'образование':
        return 'получение образования'
    
    
data['purpose_category'] = data.apply(purpose_category, axis = 1)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category purpose_category  
0                     B             None  
1                     C             None  
2                     C             None  
3                     B             None  
4                     C             None

## Ответы на вопросы.

### Вопрос 1:

Зависимость между количеством детей и возвратом кредита в срок.

In [141]:
debt_children = pd.DataFrame()
debt_children['count_dept'] = data.groupby('children')['debt'].count()
debt_children['sum_dept'] = data.groupby('children')['debt'].sum()
debt_children['result_dept'] = debt_children['sum_dept'] / debt_children['count_dept'] 
debt_children.sort_values('result_dept', ascending = False)

count_dept  sum_dept  result_dept
children                                   
4                 41         4     0.097561
2               2128       202     0.094925
1               4856       445     0.091639
3                330        27     0.081818
0              14106      1063     0.075358
5                  9         0     0.000000

In [143]:
data.pivot_table(index = 'children', values = 'debt').sort_values(by='debt', ascending = False)

debt
children          
4         0.097561
2         0.094925
1         0.091639
3         0.081818
0         0.075358
5         0.000000

In [144]:
data['children_group'] = data['children'].apply(lambda x: 'Нет детей' if x == 0 else 'Дети есть')

df_example = data.pivot_table(index = 'children_group', values = 'debt', 
                            aggfunc = ['count', 'sum', 'mean', lambda x: 1 - x.mean()])
df_example.columns = ['Кол-во пользователей', 'Кол-во должников', '% должников', '% НЕдолжников']
df_example.style.format({'% должников': '{:.2%}', '% НЕдолжников': '{:.2%}'})

#### Вывод:

Исходя из данных самыми ответственными являются многодетные заемщики и заемщики без детей. Затем в порядке возрастания детей, за исключением случая с 3-мя детьми, растет безответственность в плане погашения кредита. 

### Вопрос 2:

Зависимость между семейным положением и возвратом кредита в срок.

In [145]:
debt_family_status = pd.DataFrame()
debt_family_status['count_dept'] = data.groupby('family_status_id')['debt'].count()
debt_family_status['sum_dept'] = data.groupby('family_status_id')['debt'].sum()
debt_family_status['result_dept'] = debt_family_status['sum_dept'] / debt_family_status['count_dept'] 
debt_family_status.sort_values('result_dept', ascending = False)

count_dept  sum_dept  result_dept
family_status_id                                   
4                       2810       274     0.097509
1                       4162       388     0.093224
0                      12344       931     0.075421
3                       1195        85     0.071130
2                        959        63     0.065693

#### Вывод:

Самыми надежными заемщиками являются вдовцы и люди в разводе. Не женатые и неузаконившие свои отношения напротив меньше вызывют доверие. Женатые стабильно.

### Вопрос 3:

Зависимость между уровнем дохода и возвратом кредита в срок.

In [146]:
debt_income = pd.DataFrame()
debt_income['count_dept'] = data.groupby('total_income_category')['debt'].count()
debt_income['sum_dept'] = data.groupby('total_income_category')['debt'].sum()
debt_income['result_dept'] = debt_income['sum_dept'] / debt_income['count_dept'] 
debt_income.sort_values('result_dept', ascending = False)

count_dept  sum_dept  result_dept
total_income_category                                   
E                              22         2     0.090909
C                           16033      1360     0.084825
A                              25         2     0.080000
B                            5040       356     0.070635
D                             350        21     0.060000

#### Вывод:

Самыми ответственными заемщиками является средний класс. Затем с уровнем дохода падает и надежность заемщика.

### Вопрос 4:

Как разные цели кредита влияют на его возврат в срок.

In [147]:
debt_purpose = pd.DataFrame()
debt_purpose['count_dept'] = data.groupby('purpose_category')['debt'].count()
debt_purpose['sum_dept'] = data.groupby('purpose_category')['debt'].sum()
debt_purpose['result_dept'] = debt_purpose['sum_dept'] / debt_purpose['count_dept'] 
debt_purpose.sort_values('result_dept', ascending = False)

count_dept  sum_dept  result_dept
purpose_category                                           
оперции с автомобилем            494        42     0.085020
проведение свадьбы               793        64     0.080706
получение образования            447        32     0.071588
операции с недвижимостью         633        42     0.066351

Заемщики на недвижимость и свадьбу самые надежные, а вот автомобилисты и студенты практически с одинаковой периодичностью вызывают вопросы.

## Общий вывод:

Самые подозрительные это граждане с 2-мя и 4-мя детьми, холостые, люди с достатком ниже среднего по Москве, автомобилисты и студенты. Остальным можно смело давать кредиты. Но судя по всему лиц вызывающих подозрение - подавляющее большинство.